### DISCUSSION TITLE
CRS:84 Projection TileLayer?

### DISCUSSION THREAD

**Discussion by jfuehner** [visgl/deck.gl#5396] [Q&A]
Recently made the switch from using the Mapbox.gl/deck.gl combination to a pure deck implementation. I was previously using Mapbox purely for base maps/raster tile rendering from ArcGIS WMS services. 

I am now connecting to ArcGIS WMS services using the deck [TileLayer](https://deck.gl/docs/api-reference/geo-layers/tile-layer). The `getTileData` function is requesting tiles using the CRS:84 projection. With that said, I have noticed that when I am at max zoom and have some entities added to the map they *appear* not where they should be. 

When I start zooming in and the WMS tiles start rendering for the closer zoom level(s) the tiles and the entities start syncing/displaying at the expected position.

Previously, I was using Mapbox with the older EPSG:3857 projection which was rendering points at the higher zoom levels fine.

Could this be an issue with the service and how the imagery was published for tiles at the higher scale for CRS:84? Or is there perhaps something I should be considering on the deck side?

Any thoughts, suggestions or even speculations would be appreciated...

**Comment by kylebarron** [comment]
> The `getTileData` function is requesting tiles using the CRS:84 projection

I don't think the TileLayer currently supports the global geodetic tiling system, because it expects to display them in the OSM indexing system with Web Mercator coordinates. You'd need to overwrite your own `Tileset2D` class https://github.com/visgl/deck.gl/blob/master/modules/geo-layers/src/tile-layer/tileset-2d.js

**Reply by kylebarron** [reply]
We should probably make it easier to swap in another `tileset2D` class, so that you could make a custom subclass of tileset easily. Would you accept a PR @Pessimistress, should `tileset2D` be a prop to the layer that could be overwritten?

**Reply by jfuehner** [reply]
@kylebarron I also experience the same issue when requesting tiles with the EPSG:4326 projection as well so it does not appear to be limited to CRS:84.

**Reply by kylebarron** [reply]
`CRS:84` and `EPSG:4326` are the same thing https://gis.stackexchange.com/a/54083

The `TileLayer` currently only works with the Web Mercator projection (`EPSG:3857`)

**Comment by Pessimistress** [comment]
In 8.4.0-beta there is a new [_imageCoordinateSystem](https://github.com/visgl/deck.gl/blob/master/docs/api-reference/layers/bitmap-layer.md#_imagecoordinatesystem-number-optional) prop on the BitmapLayer.

**Reply by kylebarron** [reply]
That's true, but the TileLayer still only handles indexing in Web Mercator coordinates. So it won't locate and load tiles in other indexing systems correctly. My interpretation of the original question is to load tiles in the "global geodetic tiling system", where coordinates are in WGS84. In that indexing system, there's two tiles at zoom 0, one for each half of the globe, and then a quadtree within each half at each zoom level.

I don't expect that deck.gl support every indexing system natively, but it would be nice for advanced users to easily swap in another tileset. Right now that's not possible because `Tileset2D` isn't exported.

```js
import {Tileset2D, TileLayer} from '@deck.gl/geo-layers';

class MyTileset2D extends Tileset2D {
  // overwrite getTileIndices
}

new TileLayer({
  tileset2d: MyTileset2D
})
```
